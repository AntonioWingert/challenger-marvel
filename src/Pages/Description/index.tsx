import { useParams } from "react-router-dom";
import HeroDescription from "../../components/HeroDescription";
import { useEffect, useState } from "react";
import { MarvelApiResponse } from "../../@types/Api";
import { getCharacterById } from "../../Api";
import { Link } from "react-router-dom";

export default function Description() {
  const [heroData, setHeroData] = useState<MarvelApiResponse | null>(null);
  const { id } = useParams() as { id: string };

  useEffect(() => {
    const fetchHeroData = async () => {
      const response = await getCharacterById(id);
      if (response.data.results.length === 0) return;
      setHeroData(response);
    };
    fetchHeroData();
  }, [id]);

  const hero = heroData?.data.results[0];

  if (!hero) return (
    <>
      <div>Héroi não encontrado, por favor retorne a tela incial</div>
      <Link to='/'>Voltar</Link>
    </>
  );


  return (
    <>
      <HeroDescription id={hero?.id} name={hero?.name} thumbnail={hero?.thumbnail.path} />
    </>
  );
}