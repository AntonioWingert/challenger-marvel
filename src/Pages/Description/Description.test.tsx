import { screen } from '@testing-library/react';
import { afterEach, beforeEach, vi } from 'vitest';
import render from '../../utils/renderWithRouter';
import Description from '.';
import mockMarvelAPICharacterById from '../../mocks/HeroId.mock';


describe('<Search />', () => {
  beforeEach(() => {
    vi.spyOn(global, 'fetch').mockResolvedValue({
      json: vi.fn().mockResolvedValue(mockMarvelAPICharacterById),
    } as unknown as Response);
  });

  afterEach(() => {
    vi.resetAllMocks();
  });


  it('should render home page', async () => {
    const { container } = render(
      <Description />
    );

    const heading = await screen.findByRole('heading', {
      name: /iron man/i
    });
    const image = await screen.findByRole('img', {
      name: /Hero image Iron Man/i
    });
    const button = await screen.findByRole('link', {
      name: /voltar/i
    });

    expect(heading).toBeVisible();
    expect(image).toBeVisible();
    expect(button).toBeVisible();

    expect(container).toMatchSnapshot();
  });

  it('should render not found page with invalid hero', async () => {
    vi.spyOn(global, 'fetch').mockResolvedValue({
      json: vi.fn().mockResolvedValue({
        data: {
          results: [],
        }
      }),
    } as unknown as Response);
    const { container } = render(
      <Description />
    );

    const image = await screen.findByRole('img', {
      name: /hero not found/i
    });

    const heading = await screen.findByRole('heading', {
      name: /herói não encontrado/i
    });

    const button = await screen.findByRole('link', {
      name: /voltar/i
    });

    expect(image).toBeVisible();
    expect(heading).toBeVisible();
    expect(button).toBeVisible();
    expect(container).toMatchSnapshot();
  });

  it('should return error when api returns error', async () => {
    vi.spyOn(global, 'fetch').mockRejectedValue(
      new Error('Hero not found')
    );

    render(<Description />);

  });
});