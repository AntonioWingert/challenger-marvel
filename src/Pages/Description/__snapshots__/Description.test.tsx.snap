// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Search /> > should render home page 1`] = `
.c0 {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: space-between;
  width: 100%;
  min-height: 100vh;
  height: auto;
  background-color: #ccc;
  border-radius: 0.5rem;
}

.c0 h2 {
  font-size: 4rem;
  text-align: center;
}

.c0 img {
  width: auto;
  height: auto;
}

.c0 p {
  font-size: 1rem;
  text-align: center;
  width: 40%;
}

.c0 a {
  display: flex;
  align-items: center;
  justify-content: center;
  text-decoration: none;
  color: #F5F5F5;
  width: 35rem;
  height: 4rem;
  background-color: #4A1414;
}

@media (max-width: 320px) {

}

@media (max-width: 768px) {

}

@media (max-width: 768px) {
  .c0 a {
    width: 100%;
    height: 2rem;
  }

  .c0 img {
    width: 80%;
    height: 40vh;
  }

  .c0 p {
    width: 80%;
    text-align: justify;
  }
}

<div>
  <div
    class="c0"
  >
    <h2>
      Iron Man
    </h2>
    <img
      alt="Hero image Iron Man"
      src="http://example.com/ironman/portrait_uncanny.jpg"
    />
    <p>
      Genius. Billionaire. Playboy. Philanthropist.
    </p>
    <a
      href="/"
    >
      Voltar
    </a>
  </div>
</div>
`;

exports[`<Search /> > should render not found page with invalid hero 1`] = `
.c1 {
  font-size: 2.4rem;
  color: #0D0D0D;
  font-weight: 400;
  padding: 32px;
}

.c0 {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: space-around;
  width: 100vw;
  height: 100vh;
  background-color: #ccc;
}

.c0 a {
  display: flex;
  align-items: center;
  justify-content: center;
  text-decoration: none;
  color: #F5F5F5;
  background-color: #4A1414;
  width: 30%;
  height: 4rem;
  border-radius: 0.5rem;
  font-size: 1.5rem;
}

.c0 img {
  width: 20%;
  height: auto;
}

@media (max-width: 320px) {
  .c1 {
    font-size: 1.6rem;
  }
}

@media (max-width: 768px) {
  .c0 a {
    width: 60%;
    font-size: 1rem;
  }

  .c0 img {
    width: 80%;
    height: auto;
  }
}

<div>
  <div
    class="c0"
  >
    <img
      alt="Hero not found"
      src="/spider-man.png"
    />
    <h1
      class="c1"
    >
      Herói não encontrado
    </h1>
    <a
      href="/"
    >
      Voltar
    </a>
  </div>
</div>
`;
